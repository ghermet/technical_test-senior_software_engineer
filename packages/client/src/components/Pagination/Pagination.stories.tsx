import React, { useState } from 'react'
import '../../styles/globals.generated.css'
import Pagination from './Pagination'

export default {
  title: 'Components/Pagination',
  component: Pagination,
}

export const Default = () => {
  const [page, setPage] = useState(1)
  return <Pagination page={page} perPage={10} total={199} onPageChange={setPage} />
}
