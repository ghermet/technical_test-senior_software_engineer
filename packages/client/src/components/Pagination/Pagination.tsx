import React from 'react'
import ButtonSquare from '../ButtonSquare/ButtonSquare'
import { ChevronLeftBold, ChevronRightBold } from '@pretto/picto'
import { ctwMerge } from '../../utility/ctw-merge'

type PaginationProps = {
  className?: string
  page: number
  perPage: number
  total: number
  onPageChange: (page: number) => void
}

export const Pagination = ({ className, page, perPage, total, onPageChange }: PaginationProps) => {
  const isFirstPage = page === 1
  const isLastPage = page * perPage >= total
  const handlePrevPage = () => onPageChange(page - 1)
  const handleNextPage = () => onPageChange(page + 1)
  return (
    <div className={ctwMerge('flex flex-row items-center justify-between', className)}>
      <span>
        <span data-testid="current-page">{page}</span> -{' '}
        <span data-testid="per-page">{Math.min(page * perPage, total)}</span> of{' '}
        <span data-testid="total">{total}</span>
      </span>
      <div className="flex gap-2">
        <ButtonSquare
          onClick={handlePrevPage}
          disabled={isFirstPage}
          aria-label="Previous page"
          data-testid="desktop-prev-page-button"
        >
          <ChevronLeftBold />
        </ButtonSquare>
        <ButtonSquare
          onClick={handleNextPage}
          disabled={isLastPage}
          aria-label="Next page"
          data-testid="desktop-next-page-button"
        >
          <ChevronRightBold />
        </ButtonSquare>
      </div>
    </div>
  )
}

export default Pagination
