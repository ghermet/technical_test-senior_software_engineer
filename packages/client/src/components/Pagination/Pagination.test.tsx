import React from 'react'
import { render } from '@testing-library/react'
import Pagination from './Pagination'

describe('Pagination', () => {
  it('should render a pagination element correctly', () => {
    const { container } = render(<Pagination page={1} perPage={10} total={100} onPageChange={jest.fn} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render the current page's number", () => {
    const { getByTestId } = render(<Pagination page={1} perPage={10} total={100} onPageChange={jest.fn} />)
    expect(getByTestId('current-page').textContent).toBe('1')
  })

  it('should render the perPage items', () => {
    const { getByTestId } = render(<Pagination page={1} perPage={10} total={100} onPageChange={jest.fn} />)
    expect(getByTestId('per-page').textContent).toBe('10')
  })

  it('should render the total items', () => {
    const { getByTestId } = render(<Pagination page={1} perPage={10} total={100} onPageChange={jest.fn} />)
    expect(getByTestId('total').textContent).toBe('100')
  })

  it('should call `onPageChange` when click on the next page button', () => {
    const onPageChangeMock = jest.fn()
    const { getByTestId } = render(<Pagination page={1} perPage={10} total={100} onPageChange={onPageChangeMock} />)
    getByTestId('desktop-next-page-button').click()
    expect(onPageChangeMock).toHaveBeenCalledWith(2)
  })

  it('should call `onPageChange` when click on the previous page button', () => {
    const onPageChangeMock = jest.fn()
    const { getByTestId } = render(<Pagination page={2} perPage={10} total={100} onPageChange={onPageChangeMock} />)
    getByTestId('desktop-prev-page-button').click()
    expect(onPageChangeMock).toHaveBeenCalledWith(1)
  })

  it('should not call `onPageChange` when click on the next page button if the current page is the last one', () => {
    const onPageChangeMock = jest.fn()
    const { getByTestId } = render(<Pagination page={10} perPage={10} total={100} onPageChange={onPageChangeMock} />)
    const nextPageButton = getByTestId('desktop-next-page-button') as HTMLButtonElement
    nextPageButton.click()
    expect(onPageChangeMock).not.toHaveBeenCalled()
  })

  it('should not call `onPageChange` when click on the previous page button if the current page is the first one', () => {
    const onPageChangeMock = jest.fn()
    const { getByTestId } = render(<Pagination page={1} perPage={10} total={100} onPageChange={onPageChangeMock} />)
    const prevPageButton = getByTestId('desktop-prev-page-button') as HTMLButtonElement
    prevPageButton.click()
    expect(onPageChangeMock).not.toHaveBeenCalled()
  })

  it('should render the next page button as disabled if the current page is the last one', () => {
    const { getByTestId } = render(<Pagination page={10} perPage={10} total={100} onPageChange={jest.fn} />)
    const nextPageButton = getByTestId('desktop-next-page-button') as HTMLButtonElement
    expect(nextPageButton.disabled).toBe(true)
  })

  it('should render the previous page button as disabled if the current page is the first one', () => {
    const { getByTestId } = render(<Pagination page={1} perPage={10} total={100} onPageChange={jest.fn} />)
    const button = getByTestId('desktop-prev-page-button') as HTMLButtonElement
    expect(button.disabled).toBe(true)
  })
})
