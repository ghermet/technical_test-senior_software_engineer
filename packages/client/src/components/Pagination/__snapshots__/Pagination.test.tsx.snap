// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render a pagination element correctly 1`] = `
<div
  class="flex flex-row items-center justify-between"
>
  <span>
    <span
      data-testid="current-page"
    >
      1
    </span>
     -
     
    <span
      data-testid="per-page"
    >
      10
    </span>
     of
     
    <span
      data-testid="total"
    >
      100
    </span>
  </span>
  <div
    class="flex gap-2"
  >
    <button
      aria-label="Previous page"
      class="items-center appearance-none bg-accent1 rounded text-white cursor-pointer outline-offset-2 outline-dashed inline-flex h-8 justify-center align-middle w-8 p-0 border-none transition-colors duration-150 ease-in-out hover:bg-accent1Hover focus:outline-accent1Hover focus:outline-2 focus-visible:bg-accent1Hover focus-visible:outline-accent1Hover focus-visible:outline-2 disabled:bg-accent1Disabled disabled:cursor-not-allowed"
      data-testid="desktop-prev-page-button"
      disabled=""
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.58,2.5l1.3,1.3L6.59,8l4.29,4.2-1.3,1.3L4,8Z"
        />
      </svg>
    </button>
    <button
      aria-label="Next page"
      class="items-center appearance-none bg-accent1 rounded text-white cursor-pointer outline-offset-2 outline-dashed inline-flex h-8 justify-center align-middle w-8 p-0 border-none transition-colors duration-150 ease-in-out hover:bg-accent1Hover focus:outline-accent1Hover focus:outline-2 focus-visible:bg-accent1Hover focus-visible:outline-accent1Hover focus-visible:outline-2 disabled:bg-accent1Disabled disabled:cursor-not-allowed"
      data-testid="desktop-next-page-button"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.3,2.5,5,3.8,9.28,8,5,12.2l1.3,1.3L11.88,8Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
